### PR TITLE
[idp-1027] Support policy with custom permissions

### PR DESCRIPTION
**Pull request by PrincessMadMath**
## Description of changes
Add support for more granular and custom permissions:
- New attributes to specify required permissions
- Requirement, RequirementHandler and Policy to leverage this attribute

## Breaking changes
NO but the previous attribute has been mark as obsolete.

## Additional checks
- [X] Updated the documentation of the project to reflect the changes (not confluence yet)
- [X] Added new tests that cover the code changes

### CODE REVIEW

File: src/Workleap.AspNetCore.Authentication.ClientCredentialsGrant/AuthorizationExtensions.cs [hunk 1]
```diff
@@ -45,7 +46,16 @@ public static IServiceCollection AddClientCredentialsAuthorization(this IService
...
+                
+                authorizationOptions.AddPolicy(
+                    ClientCredentialsDefaults.AuthorizationRequirePermissionsPolicy,
+                    policy => policy
+                        .AddAuthenticationSchemes(ClientCredentialsDefaults.AuthenticationScheme)
+                        .RequireAuthenticatedUser()
+                        .AddRequirements(new RequireClientCredentialsRequirement()));
             });
+        
+        services.TryAddEnumerable(ServiceDescriptor.Singleton<IAuthorizationHandler, RequireClientCredentialsRequirementHandler>());
```
Comment (PrincessMadMath) [inline src/Workleap.AspNetCore.Authentication.ClientCredentialsGrant/AuthorizationExtensions.cs:58]: Check if TryAddEnumerable is the good method

**Review by meziantou — APPROVED** [review]

**Review by meziantou — APPROVED** [review]

File: src/Workleap.AspNetCore.Authentication.ClientCredentialsGrant/PublicAPI.Unshipped.txt [hunk 2]
```diff
@@ -1 +1,6 @@
-#nullable enable
\ No newline at end of file
+#nullable enable
```
Comment (asimmon) [inline src/Workleap.AspNetCore.Authentication.ClientCredentialsGrant/PublicAPI.Unshipped.txt:1]: to shipped.txt

File: src/Workleap.AspNetCore.Authentication.ClientCredentialsGrant/RequireClientCredentialsAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    private static readonly Dictionary<ClientCredentialsScope, string> EnumScopeNameMapping = new()
+    {
+        [ClientCredentialsScope.Read] = "read",
+        [ClientCredentialsScope.Write] = "write",
+        [ClientCredentialsScope.Admin] = "admin",
+    };
+
+    [SuppressMessage("Microsoft.Design", "CA1019:DefineAccessorsForAttributeArguments", Justification = "The arguments are transformed.")]
+    public RequireClientCredentialsAttribute(ClientCredentialsScope scope)
+        : this(EnumScopeNameMapping.GetValueOrDefault(scope) ?? throw new ArgumentException($"{scope} is not an valid scope value"))
```
Comment (asimmon) [inline src/Workleap.AspNetCore.Authentication.ClientCredentialsGrant/RequireClientCredentialsAttribute.cs:19]:
```suggestion
        : this(EnumScopeNameMapping.GetValueOrDefault(scope) ?? throw new ArgumentOutOfRangeException(nameof(scope), scope, $"'{scope}' is not an valid scope value"))
```

File: src/Workleap.Authentication.ClientCredentialsGrant.Tests/AuthorizationExtensionsTest.cs [hunk 4]
```diff
@@ -43,16 +43,37 @@ public async Task GivenIServiceCollection_WhenAddClientCredentialsAuthorization_
...
     }
+    
+    [Fact]
+    public async Task GivenIServiceCollection_WhenAddClientCredentialsAuthorization_ThenRequirementHandlerRegistered()
+    {
+        // Given
+        var services = new ServiceCollection();
 
-    private static void ValidatePolicy(AuthorizationPolicy? policy, ClientCredentialsScope scope)
+        services.AddOptions<JwtBearerOptions>(ClientCredentialsDefaults.AuthenticationScheme);
```
Comment (asimmon) [inline src/Workleap.Authentication.ClientCredentialsGrant.Tests/AuthorizationExtensionsTest.cs:64]: not required?

File: src/Workleap.Authentication.ClientCredentialsGrant.Tests/IntegrationTests.cs [hunk 5]
```diff
@@ -16,6 +16,9 @@
 
 namespace Workleap.Authentication.ClientCredentialsGrant.Tests;
 
+// TODO: Check if would be more readable + easier to evolve to:
```
Comment (asimmon) [inline src/Workleap.Authentication.ClientCredentialsGrant.Tests/IntegrationTests.cs:19]: remove it?

File: src/Workleap.AspNetCore.Authentication.ClientCredentialsGrant/RequireClientCredentialsAttribute.cs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Builder;
+
+namespace Workleap.AspNetCore.Authentication.ClientCredentialsGrant;
+
+[AttributeUsage(AttributeTargets.Class | AttributeTargets.Method)]
+public sealed class RequireClientCredentialsAttribute : AuthorizeAttribute
+{
+    /// <summary>
+    /// asd
```
Comment (asimmon) [inline src/Workleap.AspNetCore.Authentication.ClientCredentialsGrant/RequireClientCredentialsAttribute.cs:11]: Something's missing here

File: src/Workleap.AspNetCore.Authentication.ClientCredentialsGrant/RequireClientCredentialsAttribute.cs [hunk 7]
```diff
@@ -7,19 +7,46 @@ namespace Workleap.AspNetCore.Authentication.ClientCredentialsGrant;
...
+    /// </summary>
     private static readonly Dictionary<ClientCredentialsScope, string> EnumScopeNameMapping = new()
     {
         [ClientCredentialsScope.Read] = "read",
         [ClientCredentialsScope.Write] = "write",
         [ClientCredentialsScope.Admin] = "admin",
     };
 
+    /// <summary>
+    /// Verifies that the web API is called with the right classic scope. <br/>
```
Comment (asimmon) [inline src/Workleap.AspNetCore.Authentication.ClientCredentialsGrant/RequireClientCredentialsAttribute.cs:21]:
```suggestion
    /// Verifies that the endpoint is called with the right classic scope. <br/>
```

File: src/Workleap.AspNetCore.Authentication.ClientCredentialsGrant/RequireClientCredentialsAttribute.cs [hunk 8]
```diff
@@ -7,19 +7,46 @@ namespace Workleap.AspNetCore.Authentication.ClientCredentialsGrant;
...
+    /// - It will check in those claims type: scope, scp or http://schemas.microsoft.com/identity/claims/scop <br/>
```
Comment (asimmon) [inline src/Workleap.AspNetCore.Authentication.ClientCredentialsGrant/RequireClientCredentialsAttribute.cs:22]: Missing 'e' at the end of `http://schemas.microsoft.com/identity/claims/scop` (3 occurrences)

File: src/Workleap.AspNetCore.Authentication.ClientCredentialsGrant/RequireClientCredentialsAttribute.cs [hunk 9]
```diff
@@ -7,19 +7,46 @@ namespace Workleap.AspNetCore.Authentication.ClientCredentialsGrant;
...
+    /// </example>
     [SuppressMessage("Microsoft.Design", "CA1019:DefineAccessorsForAttributeArguments", Justification = "The arguments are transformed.")]
     public RequireClientCredentialsAttribute(ClientCredentialsScope scope)
-        : this(EnumScopeNameMapping.GetValueOrDefault(scope) ?? throw new ArgumentException($"{scope} is not an valid scope value"))
+        : this(EnumScopeNameMapping.GetValueOrDefault(scope) ?? throw new ArgumentOutOfRangeException(nameof(scope), scope, $"'{scope}' is not an valid scope value"))
     {
     }
     
+    /// <summary>
+    /// Verifies that the web API is called with the right granular permissions. <br/>
```
Comment (asimmon) [inline src/Workleap.AspNetCore.Authentication.ClientCredentialsGrant/RequireClientCredentialsAttribute.cs:40]:
```suggestion
    /// Verifies that the endpoint is called with the right granular permissions. <br/>
```

File: src/Workleap.Authentication.ClientCredentialsGrant.Tests/IntegrationTests.cs [hunk 10]
```diff
@@ -31,6 +28,7 @@ public IntegrationTests(ITestOutputHelper testOutputHelper)
     }
 
     [Fact]
+    [RequireClientCredentialsAttribute("asd")]
```
Comment (asimmon) [inline src/Workleap.Authentication.ClientCredentialsGrant.Tests/IntegrationTests.cs:31]: Just checking if "asd" is a typo + Attribute suffix isn't required

**Review by asimmon — APPROVED** [review]

**Pull request merged**